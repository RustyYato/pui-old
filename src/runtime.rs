//! A runtime checked identifier
//!
//! It uses an id allocator (usually a simple counter) to generate new ids,
//! then uses those ids to verifiy it's identity. You can either use the `Global`
//! allocator, which uses the largest unsigned non-zero integer possible (up to 64-bits),
//! or you can create your own id allocator using the [`make_counter`](make_counter)
//!
//! ```
//! type BackingScalar = [u8; 3];
//! # #[cfg(feature = "atomic")]
//! pui::make_counter! {
//!     pub type MyCustomIdAllocator = BackingScalar;
//! }
//! ```
//!
// You can use any primitive integer type and their non-zero variants

mod macros;
mod pool;
pub use pool::*;

/// an opaque runtime id
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct RuntimeId<T>(T);

impl<T> RuntimeId<T> {
    fn into_inner(self) -> T { self.0 }

    /// The underlying id
    pub fn get(&self) -> &T {
        // This is safe becuase the interface of `Pool*` demands that you must
        // either store a `RuntimeId`, return it, or drop it.
        //
        // In particular, the underlying value is useless because you cannot
        // convert it back to a `RuntimeId` outside of this module. This means
        // that all `RuntimeId`s must contain a unique id, and so `Pool*::take`
        // will return a unique id or `None`
        &self.0
    }
}

/// A counter that allocates new ids
///
/// # Safety
///
/// two equal ids may never exist together on the same thread
///
/// This implies that `Counter::*next()` may never repeat an id on the same thread,
/// and if it can repeat itself on different threads then `Self: !Send + !Sync`.
pub unsafe trait Counter: Copy + Eq {
    /// Get the next id, panics if there are no next ids
    fn next() -> Self;

    /// Try to get the next id, returns `None` if there are no next ids
    fn try_next() -> Option<Self>;
}

cfg_if::cfg_if! {
    if #[cfg(feature = "atomic")] {
        macro_rules! make_global {
            ($(#[$meta:meta])*) => {
                cfg_if::cfg_if! {
                    if #[cfg(feature = "nightly")] {
                        cfg_if::cfg_if! {
                            if #[cfg(target_has_atomic = "64")] {
                                crate::make_counter! {
                                    $(#[$meta])*
                                    pub type Global = core::num::NonZeroU64;
                                }
                            } else if #[cfg(target_has_atomic = "32")]{
                                crate::make_counter! {
                                    $(#[$meta])*
                                    pub type Global = core::num::NonZeroU32;
                                }
                            } else if #[cfg(target_has_atomic = "16")]{
                                crate::make_counter! {
                                    $(#[$meta])*
                                    pub type Global = core::num::NonZeroU16;
                                }
                            } else {
                                crate::make_counter! {
                                    $(#[$meta])*
                                    pub type Global = core::num::NonZeroU8;
                                }
                            }
                        }
                    } else {
                        crate::make_counter! {
                            $(#[$meta])*
                            pub type Global = core::num::NonZeroU64;
                        }
                    }
                }
            };
        }

        make_global! {
            /// A gobal allocator for runtime ids (not to be confused with a memory allocator)
            ///
            /// This can be used with [`Runtime`](super::runtime::Runtime) to easily
            /// create a new [`Runtime`](super::runtime::Runtime) [`Identifier`](super::Identifier)
            #[derive(Debug, PartialOrd, Ord, Hash)]
        }

        /// A runtime checked identifier
        ///
        /// This uses a runtime id to verify it's identity, this id is provided
        /// by the [`Counter`](Counter) trait, and ids may be reused via the [`PoolMut<C>`](PoolMut)
        pub struct Runtime<C = Global, P: PoolMut<C> = ()> {
            id: C,
            pool: P,
        }

        /// A handle to a [`Runtime`](Runtime) identifier
        #[repr(transparent)]
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
        pub struct RuntimeHandle<C = Global>(pub C);

        impl Runtime {
            /// Create a new runtime using [`Global`](Global) without reusing ids
            pub fn new() -> Self {
                Self::with_counter_and_pool(())
            }
        }
    } else {
        /// A runtime checked identifier
        ///
        /// This uses a runtime id to verify it's identity, this id is provided
        /// by the [`Counter`](Counter) trait, and ids may be reused via the [`PoolMut<C>`](PoolMut)
        pub struct Runtime<C, P: PoolMut<C> = ()> {
            id: C,
            pool: P,
        }

        /// A handle to a [`Runtime`](Runtime) identifier
        #[repr(transparent)]
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
        pub struct RuntimeHandle<C>(pub C);
    }
}

impl<C: Counter> Runtime<C, ()> {
    /// Create a new runtime using the selected `Counter` without reusing ids
    ///
    /// note: Rust will likely have a hard time inferring which counter to use
    /// so you will likely have to qualify which type to use `Runtime::<MyCounter, _>::with_counter()`
    pub fn with_counter() -> Self { Self::with_counter_and_pool(()) }

    /// Try to create a new runtime using the selected `Counter` without reusing ids
    ///
    /// If the `Counter` is exhausted, then this will return `None`. Otherwise it will
    /// return a valid instance of `Runtime`
    ///
    /// note: Rust will likely have a hard time inferring which counter to use
    /// so you will likely have to qualify which type to use `Runtime::<MyCounter, _>::with_counter()`
    pub fn try_with_counter() -> Option<Self> { Self::try_with_counter_and_pool(()) }
}

impl<C: Counter, P: PoolMut<C>> Runtime<C, P> {
    /// Create a new runtime using the selected `Counter` reusing ids with `PoolMut`
    ///
    /// note: Rust will likely have a hard time inferring which counter to use
    /// so you will likely have to qualify which type to use
    /// `Runtime::<MyCounter, _>::with_counter_and_pool(pool)`
    pub fn with_counter_and_pool(pool: P) -> Self {
        Self::try_with_counter_and_pool(pool).expect("Could not allocate a new runtime id")
    }

    /// Try to create a new runtime using the selected `Counter` reusing ids with `PoolMut`
    ///
    /// It will first try and pool an id, and if that's not possible,
    /// it will generate a new id with `Counter`. If the `Counter` is exhausted
    /// then this will return `None`. Otherwise it will return a valid instance of `Runtime`
    ///
    /// note: Rust will likely have a hard time inferring which counter to use
    /// so you will likely have to qualify which type to use
    /// `Runtime::<MyCounter, _>::with_counter_and_pool(pool)`
    pub fn try_with_counter_and_pool(mut pool: P) -> Option<Self> {
        let id = pool.take_mut().map(RuntimeId::into_inner).or_else(C::try_next)?;

        Some(Runtime { id, pool })
    }

    #[inline]
    /// A handle that this runtime identifier owns
    pub fn handle(&self) -> RuntimeHandle<C> { RuntimeHandle(self.id) }
}

unsafe impl<C: Counter> crate::Handle for RuntimeHandle<C> {}
unsafe impl<C: Counter, P: PoolMut<C>> crate::Identifier for Runtime<C, P> {
    type Handle = RuntimeHandle<C>;

    #[inline]
    fn handle(&self) -> Self::Handle { self.handle() }

    #[inline]
    fn owns(&self, handle: &Self::Handle) -> bool { self.id == handle.0 }
}

impl<C, P: PoolMut<C>> Drop for Runtime<C, P> {
    #[inline]
    fn drop(&mut self) {
        // # Safety
        //
        // here `C: Counter` -> `C: Copy` (because we only construct such `C`)
        let _ = self.pool.try_put_mut(RuntimeId(unsafe { core::ptr::read(&self.id) }));
    }
}

impl<C: Counter, P: PoolMut<C>> Eq for Runtime<C, P> {}
impl<C: Counter, P: PoolMut<C>> PartialEq for Runtime<C, P> {
    fn eq(&self, other: &Self) -> bool { self.id == other.id }
}

use core::fmt;
impl<C: fmt::Debug, P: PoolMut<C>> fmt::Debug for Runtime<C, P> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { write!(f, "Runtime({:?})", self.id) }
}
