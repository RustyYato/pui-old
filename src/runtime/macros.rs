/// Create a new type that implements [`IdAlloc`](crate::runtime::IdAlloc)
/// that can be used with [`Runtime`](crate::runtime::Runtime)
///
/// For example,
///
/// ```
/// # #[cfg(feature = "atomic")]
/// pui::make_global_id_alloc! {
///     pub type MyIdAlloc(MyId) = [u8; 3];
/// }
/// ```
///
/// will generate a 24-bit id_alloc that is 1 byte aligned. You can use any type that implements
/// [`Scalar`](crate::Scalar) as the backing type of a id_alloc.
///
/// You can then use it like so,
/// ```
/// # use pui::runtime::Runtime;
/// # #[cfg(feature = "atomic")]
/// # pui::make_global_id_alloc! { type MyIdAlloc(MyId) = [u8; 3]; }
/// # #[cfg(feature = "atomic")]
/// let runtime_id_alloc /* : Runtime<MyIdAlloc> */ = MyIdAlloc::new();
/// ```
/// or if you want to plug in a custom [`PoolMut<_>`](crate::runtime::PoolMut),
/// ```
/// # use pui::runtime::Runtime;
/// # #[cfg(feature = "atomic")]
/// # pui::make_global_id_alloc! { type MyIdAlloc(MyId) = [u8; 3]; }
/// # let pool = ();
/// # #[cfg(feature = "atomic")]
/// let runtime_id_alloc /* : Runtime<MyIdAlloc, _> */ = MyIdAlloc::with_pool(pool);
/// ```
#[macro_export]
macro_rules! make_global_id_alloc {
    ($(#[$meta:meta])*$v:vis type $name:ident($(#[$id_meta:meta])* $id:ident) = $inner:ty;) => {
        $(#[$meta])*
        $v struct $name;

        $(#[$id_meta])*
        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
        $v struct $id($inner);

        impl $name {
            /// Create a new new `Runtime`
            ///
            /// panic if the id_alloc is exhausted
            pub fn new() -> $crate::runtime::Runtime<Self> {
                $crate::runtime::Runtime::with_id_alloc(&mut Self)
            }

            /// Tryr to create a new new `Runtime`, return `None` if this id_alloc is exhausted
            pub fn try_new() -> Option<$crate::runtime::Runtime<Self>> {
                $crate::runtime::Runtime::try_with_id_alloc(&mut Self)
            }

            /// Create a new new `Runtime` with the given pool
            ///
            /// panic if the pool is empty and the id_alloc is exhausted
            pub fn with_pool<P: $crate::runtime::PoolMut<$id>>(pool: P) -> $crate::runtime::Runtime<Self, P> {
                $crate::runtime::Runtime::with_id_alloc_and_pool(&mut Self, pool)
            }

            /// Try to create a new new `Runtime` with the given pool
            /// return None if the pool is empty and the id_alloc is exhausted
            pub fn try_with_pool<P: $crate::runtime::PoolMut<$id>>(pool: P) -> Option<$crate::runtime::Runtime<Self, P>> {
                $crate::runtime::Runtime::try_with_id_alloc_and_pool(&mut Self, pool)
            }
        }
        unsafe impl $crate::runtime::IdAlloc for $name {
            type Id = $id;

            fn alloc(&mut self) -> $id {
                <Self as $crate::runtime::IdAlloc>::try_alloc(self).expect($crate::macros::concat!(
                    "Cannot overflow <",
                    $crate::macros::stringify!($name),
                    " as pui::runtime::IdAlloc>::next"
                ))
            }

            fn try_alloc(&mut self) -> Option<$id> {
                $crate::if_atomic_feature! {
                    #[allow(non_upper_case_globals)]
                    static make_runtime_NEXT_ID: <$inner as $crate::macros::Scalar>::Atomic =
                        <$inner as $crate::macros::Scalar>::INIT_ATOMIC;

                    <$inner as $crate::macros::Scalar>::inc_atomic(&make_runtime_NEXT_ID).map($id)
                }
            }
        }
    };
}

/// Create a new type that implements [`IdAlloc`](crate::runtime::IdAlloc)
/// that can be used with [`Runtime`](crate::runtime::Runtime)
/// which is implemented using a thread-local count
///
/// For example,
///
/// ```
/// pui::make_global_id_alloc_tl! {
///     pub type MyIdAlloc(MyId) = [u8; 3];
/// }
/// ```
///
/// will generate a 24-bit id_alloc that is 1 byte aligned. You can use any type that implements
/// [`Scalar`](crate::Scalar) as the backing type of a id_alloc.
///
/// You can then use it like so,
/// ```
/// # use pui::runtime::Runtime;
/// # pui::make_global_id_alloc_tl! { type MyIdAlloc(MyId) = [u8; 3]; }
/// let runtime_id_alloc /* : Runtime<MyIdAlloc> */ = MyIdAlloc::new();
/// ```
/// or if you want to plug in a custom [`PoolMut<_>`](crate::runtime::PoolMut),
/// ```
/// # use pui::runtime::Runtime;
/// # pui::make_global_id_alloc_tl! { type MyIdAlloc(MyId) = [u8; 3]; }
/// # let pool = ();
/// let runtime_id_alloc /* : Runtime<MyIdAlloc, _> */ = MyIdAlloc::with_pool(pool);
/// ```
#[macro_export]
#[cfg_attr(doc, doc(cfg(feature = "std")))]
macro_rules! make_global_id_alloc_tl {
    ($(#[$meta:meta])* $v:vis type $name:ident($(#[$id_meta:meta])* $id:ident) = $inner:ty;) => {
        $(#[$meta])*
        #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
        $v struct $name {
            thread_local: $crate::macros::PhantomData<$crate::ThreadLocal>
        }
        #[doc(hidden)]
        #[allow(non_upper_case_globals)]
        $v const $name: $name = $name { thread_local: $crate::macros::PhantomData };

        $(#[$id_meta])*
        #[derive(Clone, Copy, PartialEq, Eq)]
        $v struct $id($inner, $crate::macros::PhantomData<$crate::ThreadLocal>);

        impl $name {
            /// Create a new new `Runtime`
            ///
            /// panic if the id_alloc is exhausted
            pub fn new() -> $crate::runtime::Runtime<Self> {
                $crate::runtime::Runtime::with_id_alloc(&mut $name)
            }

            /// Tryr to create a new new `Runtime`, return `None` if this id_alloc is exhausted
            pub fn try_new() -> Option<$crate::runtime::Runtime<Self>> {
                $crate::runtime::Runtime::try_with_id_alloc(&mut $name)
            }

            /// Create a new new `Runtime` with the given pool
            ///
            /// panic if the pool is empty and the id_alloc is exhausted
            pub fn with_pool<P: $crate::runtime::PoolMut<$id>>(pool: P) -> $crate::runtime::Runtime<Self, P> {
                $crate::runtime::Runtime::with_id_alloc_and_pool(&mut $name, pool)
            }

            /// Try to create a new new `Runtime` with the given pool
            /// return None if the pool is empty and the id_alloc is exhausted
            pub fn try_with_pool<P: $crate::runtime::PoolMut<$id>>(pool: P) -> Option<$crate::runtime::Runtime<Self, P>> {
                $crate::runtime::Runtime::try_with_id_alloc_and_pool(&mut $name, pool)
            }
        }

        unsafe impl $crate::runtime::IdAlloc for $name {
            type Id = $id;

            fn alloc(&mut self) -> $id {
                <Self as $crate::runtime::IdAlloc>::try_alloc(self).expect($crate::macros::concat!(
                    "Cannot overflow <",
                    $crate::macros::stringify!($name),
                    " as pui::runtime::IdAlloc>::next"
                ))
            }

            fn try_alloc(&mut self) -> Option<$id> {
                $crate::macros::thread_local! {
                    #[allow(non_upper_case_globals)]
                    static make_runtime_NEXT_ID: $crate::macros::Cell<<$inner as $crate::macros::Scalar>::Local> =
                        $crate::macros::Cell::new(<$inner as $crate::macros::Scalar>::INIT_LOCAL);
                }

                make_runtime_NEXT_ID.with(|value| {
                    let (val, id) = <$inner as $crate::macros::Scalar>::inc_local(value.get())?;
                    value.set(val);
                    Some(id)
                }).map(|val| $id(val, $crate::macros::PhantomData))
            }
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! make_global_option_pool {
    ($(#[$meta:meta])* $v:vis one $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v struct $name;

        const _: () = {
            #[allow(non_upper_case_globals)]
            static make_global_option_pool: $crate::macros::InitFlag = $crate::macros::InitFlag::new();
            #[allow(non_upper_case_globals)]
            static mut make_global_SINGLE_REUSE_ITEM:
                $crate::macros::MaybeUninit<$crate::runtime::RuntimeId<$item>> = $crate::macros::MaybeUninit::uninit();

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    if make_global_option_pool.start_init() {
                        unsafe { make_global_SINGLE_REUSE_ITEM = $crate::macros::MaybeUninit::new(value); }
                        make_global_option_pool.finish_init();
                        Ok(())
                    } else {
                        Err(value)
                    }
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    if make_global_option_pool.start_take() {
                        let x = unsafe { make_global_SINGLE_REUSE_ITEM.as_ptr().read() };
                        make_global_option_pool.finish_take();
                        Some(x)
                    } else {
                        None
                    }
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
    ($(#[$meta:meta])* $v:vis thread_local one $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v enum $name { $name($crate::macros::PhantomData<$crate::ThreadLocal>) }
        #[allow(non_upper_case_globals)]
        const $name: $name = $name::$name($crate::macros::PhantomData);

        const _: () = {
            $crate::macros::thread_local! {
                #[allow(non_upper_case_globals)]
                static make_global_SINGLE_REUSE_ITEM:
                    $crate::macros::Cell<$crate::macros::Option<$crate::runtime::RuntimeId<$item>>> = $crate::macros::Cell::default()
            }

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    make_global_SINGLE_REUSE_ITEM.with(|c| {
                        if unsafe { (*c.as_ptr()).is_none() } {
                            c.set(Some(value));
                            Ok(())
                        } else {
                            Err(value)
                        }
                    })
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    make_global_SINGLE_REUSE_ITEM.with(|c| c.take())
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                #[inline]
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                #[inline]
                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
}

///
#[macro_export]
#[cfg(not(any(doc, feature = "std")))]
macro_rules! make_global_pool {
    ($(#[$meta:meta])* $v:vis stack $name:ident($item:ty);) => {
        $crate::macros::compile_error!{"the `std` feature on `pui` must be turned on to allow global stack pool"}
    };
    ($(#[$meta:meta])* $v:vis queue $name:ident($item:ty);) => {
        $crate::macros::compile_error!{"the `std` feature on `pui` must be turned on to allow global queue pool"}
    };
    ($(#[$meta:meta])* $v:vis one $name:ident($item:ty);) => {
        $crate::make_global_option_pool!{$(#[$meta])* $v one $name($item);}
    };
    ($(#[$meta:meta])* $v:vis thread_local stack $name:ident($item:ty);) => {
        $crate::macros::compile_error!{"the `std` feature on `pui` must be turned on to allow thread local stack pool"}
    };
    ($(#[$meta:meta])* $v:vis thread_local queue $name:ident($item:ty);) => {
        $crate::macros::compile_error!{"the `std` feature on `pui` must be turned on to allow thread local queue pool"}
    };
    ($(#[$meta:meta])* $v:vis thread_local one $name:ident($item:ty);) => {
        $crate::make_global_option_pool!{$(#[$meta])* $v thread_local one $name($item);}
    };
}

/// Create a new type that implements [`Pool`](crate::runtime::Pool) and [`PoolMut`](crate::runtime::PoolMut)
/// that can be used with [`Runtime`](crate::runtime::Runtime)
///
/// For example,
///
/// ```
/// # #[cfg(feature = "std")]
/// pui::make_global_pool! {
///     pub stack MyPool(pui::runtime::Global);
/// }
///
/// let _my_pool = MyPool;
/// ```
///
/// will generate a global pool that yields used ids in FILO order.
///
/// in place of `stack` you can also use,
///
/// * stack - FILO order
/// * thread_local stack - FILO order, but stores ids in a thread local (this is best used with thread local ids)
/// * queue - FIFO order
/// * thread_local queue - FIFO order, but stores ids in a thread local (this is best used with thread local ids)
/// * one - stores a single id, best used with a id_alloc backed by `()`
/// * thread_local one - stores a single id, best used with a thread_local id backed by `()`
///
/// in place of `pui::runtime::Global` you can use any type that implements `IdAlloc`
#[macro_export]
#[cfg(any(doc, feature = "std"))]
macro_rules! make_global_pool {
    ($(#[$meta:meta])* $v:vis stack $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v struct $name;

        const _: () = {
            #[allow(non_upper_case_globals)]
            fn make_global_get_it() -> $crate::macros::MutexGuard<'static, $crate::macros::Vec<$crate::runtime::RuntimeId<$item>>> {
                static mut make_global_REUSE: $crate::macros::MaybeUninit<$crate::macros::Mutex<$crate::macros::Vec<$crate::runtime::RuntimeId<$item>>>> =
                    $crate::macros::MaybeUninit::uninit();
                static make_global_ONCE: $crate::macros::Once = $crate::macros::Once::new();

                make_global_ONCE.call_once(|| unsafe {
                    make_global_REUSE = $crate::macros::MaybeUninit::new($crate::macros::Mutex::default());
                });

                let make_global = unsafe { &*make_global_REUSE.as_ptr() };
                make_global.lock().unwrap()
            }

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    make_global_get_it().push(value);
                    Ok(())
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    make_global_get_it().pop()
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                #[inline]
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                #[inline]
                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
    ($(#[$meta:meta])* $v:vis queue $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v struct $name;

        const _: () = {
            #[allow(non_upper_case_globals)]
            fn make_global_get_it() -> $crate::macros::MutexGuard<'static, $crate::macros::VecDeque<$crate::runtime::RuntimeId<$item>>> {
                static mut make_global_REUSE: $crate::macros::MaybeUninit<$crate::macros::Mutex<$crate::macros::VecDeque<$crate::runtime::RuntimeId<$item>>>> =
                    $crate::macros::MaybeUninit::uninit();
                static make_global_ONCE: $crate::macros::Once = $crate::macros::Once::new();

                make_global_ONCE.call_once(|| unsafe {
                    make_global_REUSE = $crate::macros::MaybeUninit::new($crate::macros::Mutex::default());
                });

                let make_global = unsafe { &*make_global_REUSE.as_ptr() };
                make_global.lock().unwrap()
            }

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    make_global_get_it().push_back(value);
                    Ok(())
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    make_global_get_it().pop_front()
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                #[inline]
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                #[inline]
                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
    ($(#[$meta:meta])* $v:vis one $name:ident($item:ty);) => {
        $crate::make_global_option_pool!{$(#[$meta])* $v one $name($item);}
    };
    ($(#[$meta:meta])* $v:vis thread_local stack $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v enum $name { $name($crate::macros::PhantomData<$crate::ThreadLocal>) }
        #[allow(non_upper_case_globals)]
        const $name: $name = $name::$name($crate::macros::PhantomData);

        const _: () = {
            $crate::macros::thread_local! {
                #[allow(non_upper_case_globals)]
                static make_global_REUSE: $crate::macros::UnsafeCell<$crate::macros::Vec<$crate::runtime::RuntimeId<$item>>> =
                    $crate::macros::UnsafeCell::default();
            }

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    make_global_REUSE.with(|pool| unsafe {
                        (&mut *pool.get()).push(value)
                    });
                    Ok(())
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    make_global_REUSE.with(|pool| unsafe {
                        (&mut *pool.get()).pop()
                    })
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                #[inline]
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                #[inline]
                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
    ($(#[$meta:meta])* $v:vis thread_local queue $name:ident($item:ty);) => {
        $(#[$meta])*
        #[derive(Clone, Copy)]
        $v enum $name { $name($crate::macros::PhantomData<$crate::ThreadLocal>) }
        #[allow(non_upper_case_globals)]
        const $name: $name = $name::$name($crate::macros::PhantomData);

        const _: () = {
            $crate::macros::thread_local! {
                #[allow(non_upper_case_globals)]
                static make_global_REUSE: $crate::macros::UnsafeCell<$crate::macros::VecDeque<$crate::runtime::RuntimeId<$item>>> =
                    $crate::macros::UnsafeCell::default();
            }

            impl $crate::runtime::PoolMut<$item> for $name {
                fn try_put_mut(&mut self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    make_global_REUSE.with(|pool| unsafe {
                        (&mut *pool.get()).push_back(value)
                    });
                    Ok(())
                }

                fn take_mut(&mut self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    make_global_REUSE.with(|pool| unsafe {
                        (&mut *pool.get()).pop_front()
                    })
                }
            }

            impl $crate::runtime::Pool<$item> for $name {
                #[inline]
                fn try_put(&self, value: $crate::runtime::RuntimeId<$item>) -> Result<(), $crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::try_put_mut(&mut $name, value)
                }

                #[inline]
                fn take(&self) -> Option<$crate::runtime::RuntimeId<$item>> {
                    <Self as $crate::runtime::PoolMut<$item>>::take_mut(&mut $name)
                }
            }
        };
    };
    ($(#[$meta:meta])* $v:vis thread_local one $name:ident($item:ty);) => {
        $crate::make_global_option_pool!{$(#[$meta])* $v thread_local one $name($item);}
    };
}
