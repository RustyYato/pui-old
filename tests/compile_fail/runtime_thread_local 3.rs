use pui::runtime::Global;

pui::make_global_pool! {
    thread_local one Foo(Global);
}

fn main() {
    fn assert_send_sync<T: Send + Sync>(_: T) {}
    assert_send_sync(Global::new());
    assert_send_sync(Global::with_pool(Foo))
}
