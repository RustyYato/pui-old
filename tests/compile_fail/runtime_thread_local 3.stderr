error[E0277]: `*mut ()` cannot be sent between threads safely
  --> $DIR/runtime_thread_local 3.rs:10:5
   |
8  |     fn assert_send_sync<T: Send + Sync>(_: T) {}
   |                            ---- required by this bound in `main::assert_send_sync`
9  |     assert_send_sync(Global::new());
10 |     assert_send_sync(Global::with_pool(Foo))
   |     ^^^^^^^^^^^^^^^^ `*mut ()` cannot be sent between threads safely
   |
   = help: within `pui::runtime::Runtime<pui::runtime::Global, Foo>`, the trait `std::marker::Send` is not implemented for `*mut ()`
   = note: required because it appears within the type `pui::ThreadLocal`
   = note: required because it appears within the type `std::marker::PhantomData<pui::ThreadLocal>`
   = note: required because it appears within the type `Foo`
   = note: required because it appears within the type `pui::runtime::Runtime<pui::runtime::Global, Foo>`

error[E0277]: `*mut ()` cannot be shared between threads safely
  --> $DIR/runtime_thread_local 3.rs:10:5
   |
8  |     fn assert_send_sync<T: Send + Sync>(_: T) {}
   |                                   ---- required by this bound in `main::assert_send_sync`
9  |     assert_send_sync(Global::new());
10 |     assert_send_sync(Global::with_pool(Foo))
   |     ^^^^^^^^^^^^^^^^ `*mut ()` cannot be shared between threads safely
   |
   = help: within `pui::runtime::Runtime<pui::runtime::Global, Foo>`, the trait `std::marker::Sync` is not implemented for `*mut ()`
   = note: required because it appears within the type `pui::ThreadLocal`
   = note: required because it appears within the type `std::marker::PhantomData<pui::ThreadLocal>`
   = note: required because it appears within the type `Foo`
   = note: required because it appears within the type `pui::runtime::Runtime<pui::runtime::Global, Foo>`
